draco/tests/test_hard.py

```diff
@@ -319,3 +320,596 @@ def test_point_tick_bar_without_x_or_y():
+    assert (
+        list_violations(
+            b.program
+            + """
+    entity(mark,root,1).
+    attribute((mark,type),1,area).
+    entity(encoding,1,2).
+    attribute((encoding,channel),2,x).
+    """
+        )
+        == ["line_area_without_x_y"]
+    )
+
+    assert (
+        list_violations(
+            b.program
+            + """
+    entity(mark,root,1).
+    attribute((mark,type),1,area).
+    entity(encoding,1,2).
+    """
+        )
+        == ["line_area_without_x_y"]
+    )
+
+
+def test_line_area_with_discrete():
+    b = hard.blocks["line_area_with_discrete"]
+    assert isinstance(b, Block)
+
+    assert no_violations(
+        b.program
+        + """
+    entity(mark,0,1).
+    entity(scale,0,2).
+    entity(scale,0,3).
+    attribute((mark,type),1,line).
+    attribute((scale,channel),2,x).
+    attribute((scale,type),2,linear).
+    attribute((scale,channel),3,y).
+    attribute((scale,type),3,linear).
+    """
+    )
+
+    assert no_violations(
+        b.program
+        + """
+    entity(mark,0,1).
+    entity(scale,0,2).
+    entity(scale,0,3).
+    attribute((mark,type),1,line).
+    attribute((scale,channel),2,x).
+    attribute((scale,type),2,categorical).
+    attribute((scale,channel),3,y).
+    attribute((scale,type),3,linear).
+    """
+    )
+
+    assert (
+        list_violations(
+            b.program
+            + """
+    entity(mark,0,1).
+    entity(scale,0,2).
+    entity(scale,0,3).
+    attribute((mark,type),1,line).
+    attribute((scale,channel),2,x).
+    attribute((scale,type),2,categorical).
+    attribute((scale,channel),3,y).
+    attribute((scale,type),3,ordinal).
+    """
+        )
+        == ["line_area_with_discrete"]
+    )
+
+
+def test_bar_tick_continuous_x_y():
+    b = hard.blocks["bar_tick_continuous_x_y"]
+    assert isinstance(b, Block)
+
+    assert no_violations(
+        b.program
+        + """
+    entity(mark,0,1).
+    attribute((mark,type),1,tick).
+    entity(encoding,1,2).
+    attribute((encoding,channel),2,x).
+    entity(scale,0,4).
+    attribute((scale,channel),4,x).
+    attribute((scale,type),4,linear).
+    """
+    )
+
+    assert no_violations(
+        b.program
+        + ("\n").join(
+            dict_to_facts(
+                {
+                    "mark": [
+                        {
+                            "type": "tick",
+                            "encoding": [
+                                {"channel": "x", "field": "temperature"},
+                                {"channel": "y", "field": "wind"},
+                            ],
+                        }
+                    ],
+                    "scale": [
+                        {"channel": "x", "type": "linear"},
+                        {"channel": "y", "type": "categorical"},
+                    ],
+                }
+            )
+        )
+    )
+
+    # multiple views
+    assert (
+        list_violations(
+            b.program
+            + ("\n").join(
```

Comment:
Since `run_clingo` actually supports taking an iterator of strings instead of a string, I wonder whether we can write `list_violations` also in a way where it supports an iterator.  Then we could write

```py
list_violations([b.program] + dict_to_facts(...))
```